<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
# Импортируем необходимые библиотеки
import pandas as pd
import psycopg2

In [2]:
# Задаем параметры соединения 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
# Создаём соединение с заданными параметрами
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# Создаем f-строку query_vac для запроса sql-таблице vacancies с вакансиями
query_vac = f'''select *
            from vacancies
         '''

df_vac = pd.read_sql_query(query_vac, connection) # Передаем результаты запроса в дата-фрейм df_vac
display(df_vac) # Распечатываем дата-фрейм df_vac для предварительного анализа

C:\Users\afina\AppData\Local\Temp\ipykernel_19248\2258313404.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac = pd.read_sql_query(query_vac, connection)


id                                               name  \
0      55312386                                Компьютерный Мастер   
1      55843014                            Системный администратор   
2      54525964                      Lead Java Developer to Poland   
3      54525965                      Lead Java Developer to Poland   
4      55354053  Специалист службы поддержки с техническими зна...   
...         ...                                                ...   
49192  55589514           Специалист по информационным технологиям   
49193  55029668                     Руководитель процессного офиса   
49194  66172263                         Начальник сервисной службы   
49195  66180086                             Senior IOS разработчик   
49196  55547842                                   DataOps Engineer   

                                              key_skills          schedule  \
0      Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1      Средства криптографической защиты информации\t...       Полный день   
2        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                                   None  Удаленная работа   
...                                                  ...               ...   
49192  Электронный документооборот\t1С: Предприятие 8...       Полный день   
49193  Оптимизация бизнес-процессов\tСтратегическое м...       Полный день   
49194  Организаторские навыки\tУправление персоналом\...       Полный день   
49195  Swift\tRealm\tiOS\tАнглийский язык\tMVP\tMVVM\...       Полный день   
49196                                               None  Удаленная работа   

               experience           employment  salary_from  salary_to  \
0               Нет опыта     Полная занятость      64000.0        NaN   
1           От 3 до 6 лет     Полная занятость          NaN        NaN   
2           От 3 до 6 лет     Полная занятость          NaN        NaN   
3           От 3 до 6 лет     Полная занятость          NaN        NaN   
4               Нет опыта  Частичная занятость      15000.0        NaN   
...                   ...                  ...          ...        ...   
49192       От 3 до 6 лет     Полная занятость          NaN        NaN   
49193       От 3 до 6 лет     Полная занятость          NaN        NaN   
49194  От 1 года до 3 лет     Полная занятость      60000.0        NaN   
49195  От 1 года до 3 лет     Полная занятость     146456.0   244094.0   
49196       От 3 до 6 лет     Полная занятость          NaN        NaN   

       area_id  employer_id  
0         1516      5724811  
1          130      4903713  
2          160        69961  
3          159        69961  
4         1955         1740  
...        ...          ...  
49192       66      5109332  
49193       26        51333  
49194     1930      4094700  
49195      160      4469939  
49196      160      4978627  

[49197 rows x 10 columns]

В базе данных vacancies - 49 197 записей (предложение вакансий от работодателей) и каждая запись состоит из 10 столбцов (каждое предложение по вакансии описывается 10 параметрами: идентификатор, название, описание навыков, описание графика работы(полный день или удаленная работа), опыт работы (нет опыта - ? лет), желаемый тип занятости(полная или частичная), нижняя вилка з/п(не у всех задана), верхняя вилка з/п(не у всех задана), идентификатор региона, идентификатор работодателя). 

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [5]:
# Создаем f-строку query_employ для запроса sql-таблице employers с работодателями
query_employ = f'''select *
            from employers
         '''

df_employ = pd.read_sql_query(query_employ, connection) # Передаем результаты запроса в дата-фрейм
                                                        # df_employ
display(df_employ) # Распечатываем дата-фрейм df_employ для предварительного анализа

C:\Users\afina\AppData\Local\Temp\ipykernel_19248\4285040227.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employ = pd.read_sql_query(query_employ, connection)


id                                               name  area
0         2393                   Программный Продукт, ИТ-компания     1
1        72977                                          БАРС Груп    88
2         3155  Бест, Торгово-производственная компания, Екате...     3
3          675                                   КОРУС Консалтинг     2
4      1840010  филиал ФКУ Налог-Сервис ФНС России в Республик...    88
...        ...                                                ...   ...
23496  1441774                                              МарКо    88
23497  4718575                                      MKomov Studio    73
23498  3702770                                  АКИБ Ипотека-банк  2759
23499  5636092                                        ГИМ-КЛИНИКА    99
23500  2064382                                             Рыббла   104

[23501 rows x 3 columns]

В базе данных employers - 23 501 записей (работодателей) и каждая запись состоит из 3 столбцов (каждый работодатель описывается 3 параметрами: идентификатор, название и идентификатор региона в котором находится работодатель).

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [6]:
# Создаем f-строку query_areas для запроса sql-таблице areas с регионами
query_areas = f'''select *
            from areas
         '''

df_areas = pd.read_sql_query(query_areas, connection) # Передаем результаты запроса в дата-фрейм
                                                      # df_areas
display(df_areas) # Распечатываем дата-фрейм df_areas для предварительного анализа

C:\Users\afina\AppData\Local\Temp\ipykernel_19248\1628710500.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_areas = pd.read_sql_query(query_areas, connection)


id              name
0     2758           Тбилиси
1        8            Майкоп
2     1180          Нерюнгри
3     1240       Новокузнецк
4        2   Санкт-Петербург
...    ...               ...
1357  5043  Звёздный Городок
1358  3934       Воротынск\n
1359  4825             Чемал
1360  3708         Усть-Нера
1361  2432         Каневская

[1362 rows x 2 columns]

В базе данных areas - 1 362 записи (региона) и каждая запись состоит из 2 столбцов (каждый регион описывается 2 параметрами: идентификатор и название).

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [7]:
# Создаем f-строку query_industr для запроса sql-таблице industries с сферами деятельности работадателей
query_industr = f'''select *
            from industries
         '''

df_industr = pd.read_sql_query(query_industr, connection) # Передаем результаты запроса в дата-фрейм
                                                          # df_industr
display(df_industr) # Распечатываем дата-фрейм df_industr для предварительного анализа

C:\Users\afina\AppData\Local\Temp\ipykernel_19248\3227769035.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_industr = pd.read_sql_query(query_industr, connection)


id                                               name
0      7.540                Разработка программного обеспечения
1      7.539  Системная интеграция,  автоматизации технологи...
2     27.550              Безалкогольные напитки (производство)
3     27.551  Безалкогольные напитки (продвижение, оптовая т...
4     13.664             Управление и эксплуатация недвижимости
..       ...                                                ...
289   37.626             Ассоциация в сфере культуры, искусства
290   42.603                              Табак  (производство)
291   29.532                           Рыболовство, рыбоводство
292   37.627                      Благотворительная организация
293  388.505  Оборудование для деревообработки, заготовки ле...

[294 rows x 2 columns]

В базе данных industries - 294 записи (сферы деятельности) и каждая запись состоит из 2 столбцов (каждая сфера деятельности описывается 2 параметрами: идентификатор и название).

In [6]:
# Создаем f-строку employers_industr для запроса sql-таблице employers_industries
employers_industr = f'''select *
            from employers_industries
         '''

df_employers_industr = pd.read_sql_query(employers_industr, connection) # Передаем результаты запроса в дата-фрейм
                                                          # df_employers_industr
df_employers_industr # Распечатываем дата-фрейм df_employers_industr для предварительного анализа

C:\Users\afina\AppData\Local\Temp\ipykernel_14488\1396660919.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employers_industr = pd.read_sql_query(employers_industr, connection) # Передаем результаты запроса в дата-фрейм


employer_id industry_id
0             2393       7.540
1             2393       7.539
2            72977       7.540
3            72977       7.539
4             3155      27.550
...            ...         ...
32328      4718575       7.539
32329      3702770      43.647
32330      2064382       7.541
32331      2064382       7.540
32332      2064382       7.539

[32333 rows x 2 columns]

В базе данных employers_industr - 32 333 записи каждая запись состоит из 2 столбцов (идентификатор работодателя и идентификатор сферы деятельности). Дополнительная таблица, которая существует для организации связи между работодателями и сферами их деятельности. Эта таблица нужна нам, поскольку у одного работодателя может быть несколько сфер деятельности (или работодатели могут вовсе не указать их). Для удобства анализа необходимо хранить запись по каждой сфере каждого работодателя в отдельной строке таблицы.

***

###### Выводы по предварительному анализу данных

Наша база вакансий состоит из 6-ти sql-таблиц, взаимосвязанных между собой. Самая информативная таблица с вакансиями - vacancies (49 197 записей), которая взвимосвязана с таблицами employers (работодатели - 23 501 записи) и  areas (регионы - 1 362 записи) через идентификаторы. В свою очередь таблица employers связана с таблицей areas и таблицей industries (сферы деятельности - 294 записи) через вспомогательную таблицу employers_industries, состоящую из идентификаторов работодателей и сфер деятельности.

Мы можем увидеть около 50 тыс. вакансий и вдвое меньшее колличество работодателей. Получается, что у одного работодателя в среднем 2 размещенных вакансии. Попробуем разобраться при дальнейшем более детальном анализе базы данных. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [4]:
'''Создаем f-строку query_4_1 для запроса sql-таблицам areas и vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_4_1 = f'''SELECT
                    a.name as areas_name,  --первый столбец таблицы будет выдавать название региона
                    COUNT(v.id) as "Количество вакансий" --второй столбец будет подсчитывать колличество вакансий в регионе
                        FROM 
                        areas a --соединяем в запросе таблицы areas и vacancies по идентификаторам региона
                        JOIN vacancies v ON v.area_id = a.id  
                        GROUP BY areas_name --группируем строки таблицы по названиям регионов
                        ORDER BY 2 DESC --сортировать будем данные по второму столбцу в порядке убывания
                        LIMIT 5 --будем выводить первые пять записей сводной таблицы
             '''

df_4_1 = pd.read_sql_query(query_4_1, connection) # Запишем результат sql-запроса query_4_1 в дата-фрейм df_4_1
display(df_4_1) # выводим на печать дата-фрейм df_4_1

C:\Users\afina\AppData\Local\Temp\ipykernel_10024\3718816395.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection) # Запишем результат sql-запроса query_4_1 в дата-фрейм df_4_1


areas_name  Количество вакансий
0           Москва                 5333
1  Санкт-Петербург                 2851
2            Минск                 2112
3      Новосибирск                 2006
4           Алматы                 1892

Как и ожидалось - лидер по колличеству размещенных вакансий - Москва и Санкт-Петербург, при чем Москва идет с значительным отрывом - около 5 тыс.вакансий, Санкт-Петербург - около 3 тыс. вакансий, Минск, Новосибирск и Алматы - около 2 тыс. вакансий.

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [4]:
'''Создаем f-строку query_4_2 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_4_2 = f'''SELECT 
                    COUNT(vacancies.id) as "Количество вакансий" --суммируем число записей, удовлетворяющих заданному критерию
                    FROM   
                    vacancies --делаем запрос к sql-таблице vacancies
                    WHERE salary_from IS NOT NULL or salary_to IS NOT NULL --отфильтровываем записи, в которых заполнено хотя бы одно из двух полей с зарплатой  
             '''
df_4_2 = pd.read_sql_query(query_4_2, connection) # Запишем результат sql-запроса query_4_2 в дата-фрейм df_4_2
display(df_4_2) # выводим на печать дата-фрейм df_4_2

C:\Users\afina\AppData\Local\Temp\ipykernel_8452\4194015420.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection) # Запишем результат sql-запроса query_4_2 в дата-фрейм df_4_2


Количество вакансий
0                24073

Колличество вакансий, где заполнено хотя бы одно из двух полей с зарплатой - 24 703, при том, что общее число записей в таблице вакансии около 50 тыс., значит у больше половины вакансий не указана вообще информация о предлагаемом уровне заработной платы. 

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [5]:
'''Создаем f-строку query_4_3 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_4_3 = f'''SELECT 
                    ROUND (AVG(salary_from)) AS "средние по нижней границе заработной платы",
                    -- в первый столбец попадет усредненное значение нижнего уровня желаемой заработной платы
                    ROUND (AVG(salary_to)) AS "средние по верхней границе заработной платы"
                    -- во второй столбец попадет усредненное значение верхнего уровня желаемой заработной платы
                    FROM 
                    vacancies --делаем запрос к sql-таблице vacancies
             '''

df_4_3 = pd.read_sql_query(query_4_3, connection) # Запишем результат sql-запроса query_4_3 в дата-фрейм df_4_3
display(df_4_3) # выводим на печать дата-фрейм df_4_3

C:\Users\afina\AppData\Local\Temp\ipykernel_8452\1346851623.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection) # Запишем результат sql-запроса query_4_3 в дата-фрейм df_4_3


средние по нижней границе заработ  средние по верхней границе зарабо
0                            71065.0                           110537.0

По размещенным вакансиям усредненный нижний уровень заработной платы - около 71 тыс. рублей, а усредненный верхний уровень заработной платы - около 110,5 тыс. рублей

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [4]:
'''Создаем f-строку query_4_4 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_4_4 = f'''SELECT 
                    vacancies.schedule as "Тип рабочего графика", --в первый столбец выводим информацию, объявленную в его названии
                    vacancies.employment as "Тип трудоустройства",--во второй столбец выводим информацию, объявленную в его названии
                    COUNT(vacancies.id) as "Количество вакансий"  --подсчитываем число вакансий по каждой группе
                    FROM   
                    vacancies --делаем запрос к sql-таблице vacancies
                    GROUP BY vacancies.schedule,vacancies.employment --группируем записи сначала по типу рабочего графика, а далее по типу трудоустройства
                    ORDER BY 3 DESC --сортируем записи по третьему столбцу в порядке убывания
                    LIMIT 2 --выводим первые две записи
             '''
df_4_4 = pd.read_sql_query(query_4_4, connection)# Запишем результат sql-запроса query_4_4 в дата-фрейм df_4_4
df_4_4 # выводим на печать дата-фрейм df_4_4

C:\Users\afina\AppData\Local\Temp\ipykernel_19728\3985129258.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)# Запишем результат sql-запроса query_4_4 в дата-фрейм df_4_4


Тип рабочего графика Тип трудоустройства  Количество вакансий
0          Полный день    Полная занятость                35367
1     Удаленная работа    Полная занятость                 7802

Наибольшее число вакансий - около 35 тыс. предлагают тип рабочего графика "полный день" и тип трудоустройства "полная знаятость". На втором месте группа из почти 8 тыс. вакансий с типом рабочего графика "удаленная работа" и типом трудоустройства "полная знаятость". Оставшиеся виды комбинаций типа рабочего графика и типа трудоустройства - получается около 7 тыс. вакансий. 

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [17]:
'''Создаем f-строку query_4_5 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_4_5 = f'''SELECT 
                    vacancies.experience as "Опыт работы", --в первый столбец выводим значения показателя "опыт работы"-experience 
                    COUNT(vacancies.id) as "Количество вакансий" --во второй столбец выводим просуммированное колличество записей по сгруппированным записям "опыт работы"-experience
                    FROM   
                    vacancies --делаем запрос к sql-таблице vacancies
                    GROUP BY vacancies.experience --группируем записи по значениям показателя "опыт работы"-experience
                    ORDER BY 2 ASC --сортируем, получившиеся значения просуммированных сгруппированных записей в порядке возрастания
                    LIMIT 4 --выводим первые четыре записи
            '''
df_4_5 = pd.read_sql_query(query_4_5, connection) # Запишем результат sql-запроса query_4_5 в дата-фрейм df_4_5
df_4_5 # выводим на печать дата-фрейм df_4_5

C:\Users\afina\AppData\Local\Temp\ipykernel_19248\248953652.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


Опыт работы  Количество вакансий
0         Более 6 лет                 1337
1           Нет опыта                 7197
2       От 3 до 6 лет                14511
3  От 1 года до 3 лет                26152

Наименьшее число записей в таблице вакансий с опытом работы более 6 лет - около 1 тыс., далее идут записи с предложением вакансий без опыта работы - около  7 тыс., опыт работы от 3-х до 6-ти лет имеют 14,5 тыс. вакансий, опыт работы от 1 года до 3-х лет имеют более половины вакансий из таблицы - 26 тыс. "Медианная группа" вакансий по показателю опыт работы это вакансии с опытом работы от 1 года до 3-х лет. Наибольшее число, размещенных вакансий для людей, проработавших по специальности более трех лет. А самые опытные специалисты с опытом работы болле 6 лет составляют самую маленькую группу. Скорее всего этих специалистов работодатели находят другим способом - "не через объявления".

***

## Выводы по детальному анализу вакансий

Чем крупнее город тем большее колличество вакансий в нем требуется. Больше половины вакансий не указывает информацию об уровне заработной платы - ни верхнюю ни нижнюю планку. Из тех вакансий, где указан желаемый уровень заработной платы - усредненный показатель составил - от 71 тыс. рублей до 110,5 тыс. рублей. Большая часть вакансий предлагает полную занятость и на полный рабочий день, хотя есть и существенная группа вакансий для работников готовых работать на удаленной работе при полной занятости (сфера АйТи этому способствует). Больщшая часть вакансий для соискателей, имеющих минимальный опыт работы - до 3-х лет, далее идет группа от 3-х до 6-ти лет, за ними следуют вакансии для людей без опыта работы и самая маленькая группа - люди с опытом работы более 6-ти лет, гипотезу  - почему эта группа самая маленькая - я писал в выводах к результатам пятого задания.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [4]:
'''Создаем f-строку query_5_1 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_5_1 = f'''(SELECT 
                    employers.name as "Название работодателя", -- в первом столбце будет как написано название работодателя
                    COUNT(vacancies.id) as "Количество вакансий" -- во втором столбце будем выводить колличество вакансий работодателя
                    FROM   
                    employers --делаем запрос к sql-таблице employers
                    JOIN vacancies ON vacancies.employer_id = employers.id --соединенной с таблицей 
                    -- vacancies по идентификатору работодатель в обоих таблицах
                    GROUP BY employers.name -- группируем по названиям работодателей
                    ORDER BY 2 DESC -- сортируем таблицу по второму столбцу по убыванию
                    LIMIT 1 -- указываем первую запись запроса
                )
                UNION ALL -- присоединяем этот же запрос, но уже с пятой записью
                (SELECT
                    employers.name as "Название работодателя",
                    COUNT(vacancies.id) as "Количество вакансий"
                    FROM   
                    employers
                    JOIN vacancies ON vacancies.employer_id = employers.id
                    GROUP BY employers.name
                    ORDER BY 2 DESC
                    OFFSET 4 LIMIT 1) -- указываем пятую запись запроса
            '''

df_5_1 = pd.read_sql_query(query_5_1, connection)  # Запишем результат sql-запроса query_5_1 в дата-фрейм df_5_1
df_5_1 # выводим на печать дата-фрейм df_5_1

C:\Users\afina\AppData\Local\Temp\ipykernel_4304\277049511.py:27: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)  # Запишем результат sql-запроса query_5_1 в дата-фрейм df_5_1


Название работодателя  Количество вакансий
0                Яндекс                 1933
1         Газпром нефть                  331

Как и ожидалось - лидерами по предлагаемым вакансиям являются крупные и известные компании.

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [5]:
'''Создаем f-строку query_5_2 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_5_2 = f'''SELECT 
                    a.name as "Название региона", -- в первом столбце показываем название региона
                    COUNT(distinct e.id) as "Количество работодателей", --суммируем колличество работодателей для этого региона без дублирования
                    COUNT(distinct v.id) as "Количество вакансий" --суммируем колличество вакансий для этого региона без дублирования
                    FROM   
                    areas a --делаем запрос к sql-таблице areas
                    JOIN employers e ON e.area = a.id --соединенной с sql-таблицей 
                    -- employers по идентификатору регион в обоих таблицах
                    LEFT JOIN vacancies v ON v.area_id = a.id -- к соединенной таблице слева присоединяем таблицу 
                    -- vacancies по идентификатору региона
                    WHERE v.id is NULL -- в агригированной таблице берем записи с отсутствием вакансий - там где отсутствует идентификатор вакансии           GROUP BY a.name
                    GROUP BY a.name -- группируем по наименованию региона
                    ORDER BY 2 DESC -- сортируем таблицу по второму столбцу по убыванию
                    LIMIT 1 -- выводим первую строку таблицу
            '''
df_5_2 = pd.read_sql_query(query_5_2, connection) # Запишем результат sql-запроса query_5_2 в дата-фрейм df_5_2
df_5_2 # выводим на печать дата-фрейм df_5_2

C:\Users\afina\AppData\Local\Temp\ipykernel_8204\1801434670.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection) # Запишем результат sql-запроса query_5_2 в дата-фрейм df_5_2


Название региона  Количество работодателей  Количество вакансий
0           Россия                       410                    0

В базе работодателей самый большой регион с отсутствием вакансий это назваение страны - Россия - это связано с тем, что сами работодатели могут работать по всей стране, но конкретные вакансии привязаны к конкретному месту в стране - городу, области, району и т.д.

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [4]:
'''Создаем f-строку query_5_3 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_5_3 = f'''SELECT 
                    e.name as "Название работодателя", -- в первом столбце показываем название работодателя
                    COUNT(distinct v.area_id) as "Количество регионов где публикуются вакансии", --суммируем колличество регионов без дублирования
                    COUNT(distinct v.id) as "Количество вакансий" --суммируем колличество вакансий без дублирования
                    FROM   
                    employers e --делаем запрос к sql-таблице employers
                    JOIN vacancies v ON v.employer_id = e.id --соединенной с таблицей 
                    -- vacancies по идентификатору работодатель в обоих таблицах
                    GROUP BY e.name -- группируем по названию работодателя
                    ORDER BY 3 DESC -- сортируем по третьему столбцу по убыванию
                    LIMIT 10 --выводим первые десять записей таблицы
             '''

df_5_3 = pd.read_sql_query (query_5_3, connection)  # Запишем результат sql-запроса query_5_3 в дата-фрейм df_5_3
df_5_3 # выводим на печать дата-фрейм df_5_3

C:\Users\afina\AppData\Local\Temp\ipykernel_19580\2293705827.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query (query_5_3, connection)  # Запишем результат sql-запроса query_5_3 в дата-фрейм df_5_3


Название работодателя  Количество регионов где публикуют  \
0                 Яндекс                                181   
1             Ростелеком                                152   
2               Тинькофф                                 43   
3                   СБЕР                                 24   
4          Газпром нефть                                 24   
5            ИК СИБИНТЕК                                 46   
6                    МТС                                 55   
7                DataArt                                  2   
8  Совкомбанк Технологии                                 26   
9             Первый Бит                                 43   

   Количество вакансий  
0                 1933  
1                  491  
2                  444  
3                  428  
4                  331  
5                  327  
6                  292  
7                  247  
8                  204  
9                  176

Лидерами по числу размещенных вакансий и охвату регионов является компания Яндекс - она идет в существенном отрыве от остальных компаний. Компания Ростелеком хотя и уступает Яндексу по колличеству вакансий, но она почти не уступает Яндексу по охвату регионов. В следующую группу я бы выделил компании: Тинькофф, СБЕР, Газпром нефть, ИК СИБИНТЕК, МТС - они охватывают около 30-ти регионов и размещают около трехсот вакансий (я округлил "на глаз"). Интересна компания Data Art - она работает в 2-х регионах, но разместила довольно большое колличество вакансий.

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [4]:
'''Создаем f-строку query_5_4 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_5_4 = f'''SELECT
                    COUNT(distinct e.id) as "Количество работодателей" -- суммируем колличество работодателей без дубляжа
                    FROM   
                    employers e --делаем запрос к sql-таблице employers
                    LEFT JOIN employers_industries i ON e.id = i.employer_id -- к таблице слева присоединяем
                    -- вспомогательную таблицу по идентификатору работодателя
                    WHERE i.industry_id IS NULL -- берем те записи, у которых не указана сфера деятельности
             '''

df_5_4 = pd.read_sql_query (query_5_4, connection) # Запишем результат sql-запроса query_5_4 в дата-фрейм df_5_4
df_5_4 # выводим на печать дата-фрейм df_5_4

C:\Users\afina\AppData\Local\Temp\ipykernel_17720\2520205357.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query (query_5_4, connection) # Запишем результат sql-запроса query_5_4 в дата-фрейм df_5_4


Количество работодателей
0                      8419

Получается, что почти у трети работодателей не указана сфера деятельности. Этот вопрос требует дополнительного изучения. Или это может крупная компания из 10-ки предыдущего задания или работодатель не счел нужным указать это в данных о себе - мелкие компании тоже могут пытаться работать в разных сферах деятельности.

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [4]:
'''Создаем f-строку query_5_5 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_5_5 = f'''SELECT
                    e.name as "Название работодателя", -- в первом столбце показываем название работодателя
                    COUNT(distinct i.industry_id) as "Количество сфер деятельности" -- во втором столбце суммируем сферы деятельности
                    FROM   
                    employers e --делаем запрос к sql-таблице employers
                    LEFT JOIN employers_industries i ON e.id = i.employer_id -- к таблице слева присоединяем
                    -- вспомогательную таблицу по идентификатору работодателя
                    GROUP BY e.name -- группируем по названию работодателя
                    HAVING COUNT(distinct i.industry_id) = 4 -- фильтруем записи только с 4-мя сферами деятельности
                    ORDER BY 1 ASC --сортируем записи по алфавиту в порядке возрастания
                    OFFSET 2 LIMIT 1 -- берем третью запись из таблицы
             '''

df_5_5 = pd.read_sql_query (query_5_5, connection)  # Запишем результат sql-запроса query_5_5 в дата-фрейм df_5_5
df_5_5 # выводим на печать дата-фрейм df_5_5

C:\Users\afina\AppData\Local\Temp\ipykernel_21624\1496312146.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query (query_5_5, connection)  # Запишем результат sql-запроса query_5_5 в дата-фрейм df_5_5


Название работодателя  Количество сфер деятельности
0                  2ГИС                             4

Результат выполнения этого задания не дает нам информацию для анализа - здесь мы просто тренируем навыки работы с sql-запросами.

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [4]:
'''Создаем f-строку query_5_6 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_5_6 = f'''SELECT
                    COUNT(distinct e.id) as "Количество работодателей" --суммируем колличество работодателей в сфере без дубляжа
                    FROM   
                    public.EMPLOYERS e --делаем запрос к sql-таблице employers  
                    LEFT JOIN public.EMPLOYERS_INDUSTRIES i ON e.id = i.employer_id -- к таблице слева присоединяем
                    -- вспомогательную таблицу по идентификатору работодателя
                    JOIN public.INDUSTRIES y ON y.id = i.industry_id
                    --присоединяем еще третью таблицу со сферами деятельности через иднтификатор сферы деятельности
                    WHERE y.name = 'Разработка программного обеспечения' --отфильтровуем нужную нам сферу деятельности
             '''

df_5_6 = pd.read_sql_query (query_5_6, connection)  # Запишем результат sql-запроса query_5_6 в дата-фрейм df_5_6
df_5_6 # выводим на печать дата-фрейм df_5_6

C:\Users\afina\AppData\Local\Temp\ipykernel_19864\1682323519.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query (query_5_6, connection)  # Запишем результат sql-запроса query_5_6 в дата-фрейм df_5_6


Количество работодателей
0                      3553

Примерно 15 % работодателей из базы работают в сфере 'Разработка программного обеспечения'

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [4]:
# Данный код необходим нам для создания списка городов-миллионников mln_cities

from urllib.parse import quote_plus
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
mln_cities = tables[1]['Город'].values.tolist()

In [5]:
mln_cities

['Саратов', 'Тюмень']

In [6]:
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) + req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
mln_cities_2 = tuple(pd.read_html(url)[1]['Город'])

Запрос перестал работать и я создал спсок городов-миллионников вручную.

In [4]:
mln_cities_2 = ('Саратов', 'Тюмень', 'Москва', 'Тюмень', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
mln_cities_2

('Саратов',
 'Тюмень',
 'Москва',
 'Тюмень',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [5]:
'''Создаем f-строку query_5_7 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_5_7 = f'''
                SELECT
                    a.name as region, --в первом столбце пишем название региона
                    COUNT(*) as vacancies_count --суммируем колличество вакансий для данного региона
                    FROM
                    vacancies as v --делаем запрос к sql-таблице vacancies
                    JOIN employers as e ON v.employer_id = e.id --соединяем ее с таблицей работодателей через идентификатор работодателя
                    JOIN areas as a ON v.area_id = a.id -- и с таблицей регионы через идентификатор регионов
                    GROUP BY
                    v.area_id, a.name, e.id, e.name -- группируем записи последовательно по схеме
                    -- идентификатор региона-название региона-идентификатор работодателя- название работодателя
                    HAVING
                    a.name IN {tuple(mln_cities_2)} AND e.name = 'Яндекс'--фильтруем таблицу по названию региона из списка городов-миллионников и названия работодаителя - Яндекс
    
                UNION ALL -- присоединяем ниже строку Total с общим количеством вакансий компании Яндекс 
                SELECT
                    'Total', -- первый столбец записи называем - Total
                    SUM(x.cnt) -- передаем просуммированное общее колличество вакансий из запроса
                FROM
                (
                    SELECT
                        COUNT(*) as cnt --суммируем число вакансий
                        FROM
                        vacancies as v --делаем запрос к sql-таблице vacancies
                        JOIN employers as e ON v.employer_id = e.id --соединяем ее с таблицей работодателей через идентификатор работодателя
                        JOIN areas as a ON v.area_id = a.id -- и с таблицей регионы через идентификатор регионов
                        GROUP BY
                        v.area_id, a.name, e.id, e.name -- группируем записи последовательно по схеме
                        -- идентификатор региона-название региона-идентификатор работодателя- название работодателя
                        HAVING
                        a.name IN {tuple(mln_cities_2)} AND e.name = 'Яндекс'--фильтруем таблицу по названию региона из списка городов-миллионников и названия работодаителя - Яндекс
                ) AS x -- результат второго запроса называем переменной x
        
                ORDER BY 2 -- полученную таблицу сортируем по колличеству вакансий
            '''

df_5_7 = pd.read_sql_query (query_5_7, connection) # Запишем результат sql-запроса query_5_7 в дата-фрейм df_5_7
df_5_7  # выводим на печать дата-фрейм df_5_7

C:\Users\afina\AppData\Local\Temp\ipykernel_19212\3201569354.py:40: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query (query_5_7, connection) # Запишем результат sql-запроса query_5_7 в дата-фрейм df_5_7


region  vacancies_count
0            Тюмень             20.0
1              Омск             21.0
2         Челябинск             22.0
3           Саратов             23.0
4        Красноярск             23.0
5         Волгоград             24.0
6    Ростов-на-Дону             25.0
7            Казань             25.0
8             Пермь             25.0
9               Уфа             26.0
10           Самара             26.0
11        Краснодар             30.0
12          Воронеж             32.0
13      Новосибирск             35.0
14  Нижний Новгород             36.0
15     Екатеринбург             39.0
16  Санкт-Петербург             42.0
17           Москва             54.0
18            Total            528.0

Мы видим, что компания Яндекс разместила предложения о найме работникам во всех городах-миллионниках. Самое большое число размещенных вакансий в Москве, далее Санкт-Петербург и т.д. Чем крупнее город - тем больше вакансий - вполне ожидаемая закономерность.

***

# Выводы по анализу работодателей

Лидером по размещенным вакансиям в данной базе является компания Яндекс - она охватывает самое большое колличество регионов и разместила больше всего вакансий. Она такжк представлена во всех крупных городах России. Около 15 % работодателей работают в сфере "Разработка програмного обеспечения".

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [4]:
'''Создаем f-строку query_6_1 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_6_1 = '''SELECT
                    COUNT(*) --считаем число записей в отфильтрованной таблице
                    FROM vacancies --делаем запрос к sql-таблице vacancies
                    WHERE name ILIKE '%data%' OR name ILIKE '%данн%' -- с помощью оператора ILIKE отбираем вакансии, в тексте описания которых присутствуют корни data- или данн-
            '''

df_6_1 = pd.read_sql_query (query_6_1, connection) # Запишем результат sql-запроса query_6_1 в дата-фрейм df_6_1
df_6_1 # выводим на печать дата-фрейм df_6_1

C:\Users\afina\AppData\Local\Temp\ipykernel_17628\1513116677.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query (query_6_1, connection) # Запишем результат sql-запроса query_6_1 в дата-фрейм df_6_1


count
0   1771

Проводя лексическую обработку названия вакансии мы видим, что из 50 тыс. вакансий в базе почти 2 тыс. относятся к работе с данными. 

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [16]:
'''Создаем f-строку query_6_2 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_6_2 = '''SELECT
                    COUNT(*) --считаем число записей в отфильтрованной таблице
                    FROM
                    vacancies v --делаем запрос к sql-таблице vacancies
                    WHERE -- проводим фильтрацию строк по названию вакансий
                    -- а также по указанию в столбце опыт - 'Нет опыта', занятость - 'стажировка'
                    (
                     v.name ILIKE '%data scientist%' OR
                     v.name ILIKE '%data science%' OR
                     v.name ILIKE '%исследователь данных%' OR
                     (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
                     v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%'
                     )
                    AND
                    (
                     v.name ILIKE '%junior%' OR
                     v.experience = 'Нет опыта' OR
                     v.employment ILIKE '%стажировка%'
                    )
                '''

df_6_2 = pd.read_sql_query (query_6_2, connection) # Запишем результат sql-запроса query_6_2 в дата-фрейм df_6_2
df_6_2 # выводим на печать дата-фрейм df_6_2

C:\Users\afina\AppData\Local\Temp\ipykernel_11360\2622205437.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query (query_6_2, connection)


count
0     51

Проводя лексическую обработку столбцов, описывающих вакансии мы видим, что из 50 тыс. вакансий в базе 51 - для начинающего дата-сайентиста. 

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [4]:
'''Создаем f-строку query_6_3 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_6_3 = f'''
                SELECT
                    COUNT (vacancies.name) as n_vac --считаем число записей в отфильтрованной таблице
                    FROM vacancies --делаем запрос к sql-таблице vacancies
                    WHERE
                    -- проводим фильтрацию вакансий по названиям, связанным с работой с данными
                    (lower(vacancies.name) LIKE '%data scientist%' 
                    OR lower(vacancies.name) LIKE '%data science%' 
                    OR lower(vacancies.name) LIKE '%исследователь данных%' 
                    OR lower(vacancies.name) LIKE '%ml%' and vacancies.name NOT LIKE '%HTML%' and vacancies.name not like '%QML%' 
                    OR lower(vacancies.name) LIKE '%machine learning%' 
                    OR lower(vacancies.name) LIKE '%машинн%обучен%') 
                    AND
                    -- а также дополнительно фильтруем навыки в записях, связанные с SQL или postgres
                    (lower(vacancies.key_skills) LIKE '%postgres%' OR
                     lower(vacancies.key_skills) LIKE '%sql%') 
              '''
df_6_3 = pd.read_sql_query (query_6_3, connection) # Запишем результат sql-запроса query_6_3 в дата-фрейм df_6_3
df_6_3 # выводим на печать дата-фрейм df_6_3

C:\Users\afina\AppData\Local\Temp\ipykernel_19304\2206134256.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query (query_6_3, connection) # Запишем результат sql-запроса query_6_3 в дата-фрейм df_6_3


n_vac
0    229

Из почти 2 тыс. вакансий, связанных с работой с данными около 200 имеют навыки работы с SQL или postgres

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [26]:
'''Создаем f-строку query_6_4 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_6_4 = f'''
                SELECT
                    COUNT(*) --считаем число записей в отфильтрованной таблице
                    FROM
                    vacancies v --делаем запрос к sql-таблице vacancies
                    WHERE
                    -- проводим фильтрацию вакансий по названиям, связанным с работой с данными
                    -- а также - присутствие в опыте вакансии упоминания о Python 
                    (
                    name ILIKE '%data scientist%'  OR
                    name ILIKE  '%data science%' OR
                    name ILIKE  '%исследователь данных%'  OR
                    (
                    name ILIKE  '%ML%' AND NAME NOT ILIKE '%HTML%')  OR
                    name ILIKE '%machine learning%' OR  
                    name ILIKE '%машинн%обучен%'
                    )
                    and
                    key_skills ILIKE '%Python%'
                '''

df_6_4 = pd.read_sql_query (query_6_4, connection) # Запишем результат sql-запроса query_6_4 в дата-фрейм df_6_4
df_6_4 # выводим на печать дата-фрейм df_6_4

C:\Users\afina\AppData\Local\Temp\ipykernel_11360\3087724236.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query (query_6_4, connection)


count
0    357

Из 1700 вакансий для около 20 % требуется навыки работы с Python.

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [31]:
'''Создаем f-строку query_6_5 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_6_5 = f'''
                SELECT 
                ROUND(AVG(LENGTH(key_skills)-LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2) as result
                -- выводим результат обработки в строках столбца key_skills 
                -- из длины исходной строки вычесть длину строки, в которой все вхождения этого символа заменены на пустую строку ('')
                -- Ключевые навыки разделены символами табуляции. В SQL этот символ обозначается как CHR(9).
                -- Число символов табуляции + 1 говорит о числе навыков в вакансии
                -- Результат такой обработки строк мы округляем до 2-х знаков после запятой
                FROM 
                vacancies v --делаем запрос к sql-таблице vacancies
                WHERE
                -- проводим фильтрацию вакансий по названиям, связанным с работой с данными
                (
                  name ILIKE '%data scientist%' OR
                  name ILIKE '%data science%' OR
                  name ILIKE '%исследователь данных%' OR
                (
                  name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
                  name ILIKE '%machine learning%' OR
                  name ILIKE '%машинн%обучен%'
                )
             '''

df_6_5 = pd.read_sql_query (query_6_5, connection) # Запишем результат sql-запроса query_6_5 в дата-фрейм df_6_5
df_6_5 # выводим на печать дата-фрейм df_6_5

C:\Users\afina\AppData\Local\Temp\ipykernel_11360\1682566586.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query (query_6_5, connection)


result
0    6.41

В вреднем по вакансиям в DS требуется 6,41 навык.

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [36]:
'''Создаем f-строку query_6_6 для запроса sql-таблице vacancies для получения необходимого
дата-фрейма, который сможет ответить на вопрос задания '''

query_6_6 = f'''
                SELECT 
                    experience as "опыт работы",
                    ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0)
                    as "средний размер зарплаты"
                    -- вычисляем средний размер заработной платы по записям, как усредненный показатель
                    -- среднего арифметического между верхним и нижним размером заработной платы
                    -- результат округляем до целого
                    FROM vacancies v --делаем запрос к sql-таблице vacancies
                    WHERE
                    -- проводим фильтрацию вакансий по названиям, связанным с работой с данными
                    -- опыт работы 'От 3 до 6 лет'
                    -- столбцы с уровнями заработной платы от и до должны быть заполнены
                    (
                    name ILIKE '%data scientist%' OR
                    name ILIKE '%data science%' OR
                    name ILIKE '%исследователь данных%' OR
                    (
                    name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
                    name ILIKE '%machine learning%' OR
                    name ILIKE '%машинн%обучен%'
                    )
                    AND experience = 'От 3 до 6 лет'
                    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                    GROUP BY experience
                '''

df_6_6 = pd.read_sql_query (query_6_6, connection) # Запишем результат sql-запроса query_6_6 в дата-фрейм df_6_6
df_6_6 # выводим на печать дата-фрейм df_6_5

C:\Users\afina\AppData\Local\Temp\ipykernel_11360\744937547.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query (query_6_6, connection)


опыт работы  средний размер зарплаты
0  От 3 до 6 лет                 243115.0

Получаем, что для специалистов DS с опытом работы от 3 до 6 лет средний уровень заработной платы составляет около 240 тыс. рублей. - довольно высокий уровень.

## Выводы по предметному анализу

Проанализировав вакансии в основном метом лексического анализа мы видим, что специалисты DS занимают в базе существенную долю, навыки Python и SQL тоже имеет среди них значительную долю. Уровень заработной платы у таких специалистов довольно высокий. 

_______________________________________________________________________________________________

# Юнит 7. Самостоятельные запросы к базе

Далее я решил сделать два запроса по базе - поизучать зависимости - между работодателями и числом вакансий, между работодателями и регионами. 

In [4]:
query_7_1 = f'''
SELECT 
e.name name_empl, --название работодателя
v.name name_vac   --название вакансии
FROM vacancies v --делаем запрос к sql-таблице vacancies
JOIN employers e ON v.employer_id = e.id --эта таблицу соединяю с таблицей employers через идентификатор работодателя
GROUP BY e.name,v.name --группирую по названиям работодателя и вакансии
ORDER BY 1,2 --сортирую по алфавиту по двум столбцам
             '''
df_7_1 = pd.read_sql_query (query_7_1, connection)
df_7_1


C:\Users\afina\AppData\Local\Temp\ipykernel_6436\1377140847.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_1 = pd.read_sql_query (query_7_1, connection)


name_empl  \
0      1000 мелочей (ИП Гафиатуллина Гульшат Митарисо...   
1                                              1001 Soft   
2                                               1001 Тур   
3                                              100 печей   
4                                                   1019   
...                                                  ...   
36517                                                ЯСП   
36518                                                ЯСП   
36519                                                ЯСП   
36520                                  Ястро-Переработка   
36521                                               ЯТСС   

                                   name_vac  
0                          Специалист по IT  
1                   Системный администратор  
2                          Perl разработчик  
3                            Программист 1С  
4                   Системный администратор  
...                                     ...  
36517      Frontend-разработчик / Fullstack  
36518                      Java-разработчик  
36519               Junior Java-разработчик  
36520               Системный администратор  
36521  Инженер службы технической поддержки  

[36522 rows x 2 columns]

Мы видим, что в таблице дублируются названия работодателей.

In [5]:
df_7_1["name_empl"].nunique()

14766

Из 36522 записей в таблице уникальных 14766 названия, значит на некоторых работодателей приходится несколько вакансий.

In [6]:
unique_values2 = df_7_1["name_empl"].value_counts()
unique_values2

name_empl
СБЕР                                                     322
МТС                                                      247
ИК СИБИНТЕК                                              237
Ростелеком                                               218
Газпром нефть                                            194
                                                        ... 
Гурьева Мария Николаевна                                   1
ГУСП МТС Центральная РБ                                    1
ГУ Тульской Области Организатор Перевозок и Навигации      1
ГУФССП России по Кемеровской области - Кузбассу            1
ЯТСС                                                       1
Name: count, Length: 14766, dtype: int64

Выведим по каждому работодателю число размещенных им вакансий.

In [8]:
unique_values2.head(15)

name_empl
СБЕР                       322
МТС                        247
ИК СИБИНТЕК                237
Ростелеком                 218
Газпром нефть              194
Тинькофф                   163
VK                         129
Яндекс                     127
Совкомбанк Технологии      120
DataArt                    106
Ozon                       100
Первый Бит                  94
Лига Цифровой Экономики     87
МАГНИТ, Розничная сеть      79
Банк ВТБ (ПАО)              76
Name: count, dtype: int64

Первые 11 работодателей разместили более 100 вакансий.

__________________________________________________________________________________________

Аналогично предыдущему запросу делаю таблицу связи работодателей и регионов в которых они разместили вакансии.

In [4]:
query_7_2 = f'''
SELECT
a.name area_name,
e.name employ_name
FROM employers e 
JOIN areas a ON e.area = a.id
GROUP BY a.name,e.name
             '''
df_7_2 = pd.read_sql_query (query_7_2, connection)
df_7_2


C:\Users\afina\AppData\Local\Temp\ipykernel_18952\3455520739.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_7_2 = pd.read_sql_query (query_7_2, connection)


area_name                        employ_name
0               Россия                         Коттон Вэй
1      Санкт-Петербург                       UNISTORY LLC
2               Тюмень                Менделеев маркетинг
3               Атырау                       ЗаманКвантор
4      Санкт-Петербург            Гефест, группа компаний
...                ...                                ...
23465            Минск                      Новаком Групп
23466           Москва                           Тематика
23467        Швейцария            Assaia International AG
23468     Новороссийск  Центр винного туризма Абрау Дюрсо
23469           Москва                               Криз

[23470 rows x 2 columns]

In [5]:
df_7_2["area_name"].nunique()

654

Из 23740 регионов 654 уникальных, значит на некоторые регионы приходится несколько работодателей.

In [7]:
unique_values = df_7_2["area_name"].value_counts()
unique_values

area_name
Москва                              5843
Санкт-Петербург                     2211
Минск                               1114
Алматы                               721
Екатеринбург                         609
                                    ... 
Каневская                              1
Железногорск (Красноярский край)       1
Гаврилов-Ям                            1
Дубовое                                1
Мама                                   1
Name: count, Length: 654, dtype: int64

In [18]:
unique_values.head(40)

area_name
Москва                       5843
Санкт-Петербург              2211
Минск                        1114
Алматы                        721
Екатеринбург                  609
Новосибирск                   571
Казань                        480
Нижний Новгород               426
Россия                        410
Краснодар                     409
Ростов-на-Дону                335
Красноярск                    331
Ташкент                       326
Самара                        314
Нур-Султан                    312
Челябинск                     310
Воронеж                       292
Уфа                           252
Пермь                         250
Тюмень                        218
Омск                          210
Казахстан                     207
Томск                         191
Ижевск                        189
Иркутск                       187
Барнаул                       187
Калининград                   157
Тула                          149
Владивосток                   148
Сара

Первые 40 регионов насчитывают более 100 работодателей. Лидеры в таблице естественно Москва и Санкт-Петербург.

***

# Общий вывод по проекту

Данный проект является хорошим тренажером для отработки навыков по работе в связке инструментов SQL и Python.

В начале мы поочередно делаем предварительный просмотр SQL-таблиц: смотрим какие признаки(столбцы) у каждой таблицы, ее размер (колличество записей), взаимосвязи между таблицами через идентификаторы (желательно прочертить схему взаимосвязей между таблицами на бумаге - это даст нам понимание каким образом можно дальше строить запросы.

Дальше мы будем изучать те записи в базе (таблицы), которые имеют наибольшее колличество признаков импортирую записи из таблиц в дата-фрейм для дальнейшей его обработки инструментами библиотек Python.
Делая запросы в SQL мы используем инструменты соединения таблиц через идентификаторы, агрегирующие функции, фильтры значений , сортировка и прочее. Полезными в дальнейшей работе будут приемы работы из раздела 6, которые дают нам возможность делать лексический анализ базы и делать из него выводы.

Приемы, отработанные в данном проекте можно применять в будующей моей работе, как специалиста DS для анализа уже более сложных баз данных.